doc/compiler/optimizer/IntroLoopOptimizations.md

```diff
@@ -287,3 +287,80 @@ the outer loop decide where to send it. It will send it to the header of the slo
+`10` 100% times when it is profiled, when it comes to the next compilation, we look at the profiling information for
+`n`, and we could check outside the loop: if `n == 10` (the profiled value). Before it goes into the loop, it checks if
+`i < 10` instead of `i < n`. Otherwise, if `n` is not equal to `10`, we go into another version of the loop, the original
+version of the loop that checks `i < n`.
+
+### Loop Versioner & Loop Specializer
+
+Loop Versioner runs before Loop Specializer. As mentioned in section 2, Loop Versioner creates tests corresponding to
+exception checks. There are a fast version of the loop and a slow version of the loop. The fast version has all the checks
+that could be removed have been removed. The slow version has all the checks intact. The Loop Specializer does not do
+anything to the slow version of the loop. Loop Specializer specializes the fast version of the loop.
+
+After both Loop Versioner and Loop Specializer have run, there are three versions of the loop. (1) One version is marked
+as cold with all the checks, and nothing is changed to constants. (2) Another version of the loop has all the checks that
+could be removed have been removed, and all the expressions that could be transformed into constants have been transformed.
+(3) The third version of the loop is somewhere in between. This intermediate loop is versioned but not specialized. It has
+all the checks that could be removed have been removed, but the constants are not specialized or substituted in the loop.
+
+### Reasons That Loop Specializer Is Not Combined With Loop Versioner
+
+Profiling information is not as strong or guaranteed as exception checks are. The exception checks should not fail in well
+behaved programs. However, values or expressions could change in well behaved programs.  In Loop Versioner or Loop Specializer,
+the tests are all or nothing. Even if one condition fails in the loop pre-header, it goes to the slow version of the loop
+or the less optimized version.
+
+For example, if a version that has `50` exception checks and one profiling condition and the profiled value changes, that
+one condition failing sends the control to the slow version of the loop where there are `50` exception checks, and that
+one value is not a constant. It is a too high of price to pay. Therefore, it is not preferred to combine the extremely
+high probability tests with nearly high probability tests. They should be separated. The extremely high probability tests
+should remain in one version. Nearly high probability tests should have their version of the loop. That is why there are
+three versions of the loop.
+
+### Similarities to Loop Versioner
+
+For example, there is an expression `x.f.g` that is profiled, and it is loop invariant. If there is a check `x.f.g == 10`
+in the loop, the expression can be hoisted to the loop pre-header. If `x.f.g` equals to `10`, it goes to the fast version
```

Comment:
> it goes to the fast version

We should replace "it" with "execution".

Reply:
Addressed in c545174